torchdata/dataloader2/reading_service.py

```diff
@@ -270,13 +286,24 @@ def initialize(self, datapipe: DataPipe) -> DataPipe:
             local_datapipe = communication.iter.QueueWrapper(
                 communication.protocol.IterDataPipeQueueProtocolClient(req_queue, res_queue)
             )
-            self.datapipes.append(local_datapipe)
+            self._worker_datapipes.append(local_datapipe)
+
+        end_datapipe = communication.iter._IterateQueueDataPipes(self._worker_datapipes)  # type: ignore[assignment]
+        self._worker_consumer_datapipe = end_datapipe
 
-        self.end_datapipe = communication.iter._IterateQueueDataPipes(self.datapipes)  # type: ignore[assignment]
         if self.main_prefetch_cnt > 0:
-            self.end_datapipe = self.end_datapipe.prefetch(self.main_prefetch_cnt)  # type: ignore[union-attr]
+            end_datapipe = self._worker_consumer_datapipe.prefetch(self.main_prefetch_cnt)  # type: ignore[union-attr]
+            self._main_prefetch_datapipe = end_datapipe
+
+        # Attach non-replicable DataPipes
+        if replicable_dp is not datapipe:
```

Comment:
I guess this happens when we have non-replicable data pipe (and need to keep it in the main process). 

And the reason to replace `replicable_dp` with `end_datapipe` is because `end_datapipe` has "exchange sink" attached? (i.e. the `_IterateQueueDataPipes` on line 291)?

Reply:
Doing that because we send `replicable_dp` to worker process, which returns an `end_datapipe` with the exchange sink.
```
non_rep_dp1 (datapipe) -> non_rep_dp2 -> rep_dp_1 (replicable_dp) -> rep_dp2 -> ...
```
After replacement, the graph becomes
```
non_rep_dp1 (datapipe) -> non_rep_dp2 -> end_datapipe
```
In this case, we want to return `non_rep_dp1` (`datapipe`) rather than the `end_datapipe`.

For the case that the whole graph is replicable
```
rep_dp_1 (replicable_dp/datapipe) -> rep_dp2 -> ...
```
After launching mp, graph becomes
```
end_datapipe
```
So, we just need to keep `end_datapipe`.